Upgrade to Hibernate Search 6.0.0.Beta9 + Elasticsearch 7.8

yrodiere opened the pull request:
Relevant: https://in.relation.to/2020/08/03/hibernate-search-6-0-0-Beta9/

yrodiere committed:
Upgrade to Hibernate Search 6.0.0.Beta9

yrodiere committed:
Upgrade to Elasticsearch 7.8.0

yrodiere committed:
hsearch-es: Use HSearch's native concept of default (unnamed) backend instead of defining our own

yrodiere committed:
hsearch-es: Remove the ability to define a *named*, default backend

This feature probably wasn't used a lot by Quarkus users, since it's
only (potentially) useful when you have multiple backends.

yrodiere committed:
hsearch-es: Add missing 'final' keyword

yrodiere committed:
hsearch-es: Move to HSearch's new simplified syntax for specifying default configuration for indexes

gsmet commented on extensions/hibernate-search-elasticsearch/deployment/src/main/java/io/quarkus/hibernate/search/elasticsearch/HibernateSearchElasticsearchProcessor.java:
```diff
@@ -97,37 +97,34 @@ void setRuntimeConfig(HibernateSearchElasticsearchRecorder recorder,
+            // There is at least one named backend: we disallow using the default backend.
+            // Theoretically we could allow using it,
+            // but then we would be unable to detect whether the default backend is used
+            // (and thus absolutely needs the version to be configured)
```
Not sure I follow you here. We make the version mandatory for named backend so maybe we should just make it mandatory for everything, default backend included? And then we could keep the named backend active?

yrodiere replied:
If we do that, then people using named backends exclusively (which, IMO, is what they should do) will have to set the version for the default backend even if they don't use it. I can do that if you want, but I'm not sure it's a better solution.

yrodiere replied:
I fixed this according to our discussion on Zulip, rebased and force-pushed. Let's wait for CI, but then as far as I know it can be merged.

gsmet commented on extensions/hibernate-search-elasticsearch/runtime/src/main/java/io/quarkus/hibernate/search/elasticsearch/runtime/HibernateSearchElasticsearchRuntimeConfig.java:
```diff
@@ -128,9 +128,18 @@
+         */
+        @ConfigItem(name = "index-defaults")
+        @Deprecated
+        ElasticsearchIndexConfig legacyIndexDefaults;
```
Given we are still experimental, I would just drop it and document it in the release notes.

yrodiere replied:
Ok, I'll add a commit.

gsmet reviewed (commented):
I added some question/suggestion. The first one is a real question: I'm not sure I understand why the default backend now needs to be disabled?

yrodiere commented:
I had a quick look at the build failure, and it seems unrelated. I suspect a problem on master?

yrodiere committed:
hsearch-es: Drop the deprecated '.index-defaults.foo' syntax for specifying default configuration for indexes

yrodiere commented:
@gsmet Also, for the migration guide:

``````
## Hibernate Search + Elasticsearch (Preview)

* Configuration properties for index defaults can no longer include the `.index-defaults` component. For example,  `quarkus.hibernate-search.elasticsearch.index-defaults.schema-management.required-status` should now be written simply as `quarkus.hibernate-search.elasticsearch.schema-management.required-status`.
* `query.explain()` now expects to be passed an entity ID instead of a document ID. In short, if your entity has an ID of type `Long`, you need to pass a `Long` (previously you had to pass a `String`).
* `@IndexedEmbedded.storage`/`ObjectFieldStorage` have been renamed to `@IndexedEmbedded.structure`/`ObjectStructure`. The older syntax is deprecated and will eventually be removed.
* `quarkus.hibernate-search.elasticsearch.default-backend` no longer exists. You can either configure a default backend (`quarkus.hibernate-search.elasticsearch.someProperty someValue`) or named backends (`quarkus.hibernate-search.elasticsearch.backends."backend-name".someProperty someValue`), but a named backend can no longer be considered as the default backend.
``````

And depending on what we decide:

```
* The default backend (configured with `quarkus.hibernate-search.elasticsearch.someProperty someValue`) can no longer be used if you define named backends (configured with `quarkus.hibernate-search.elasticsearch.backends."backend-name".someProperty someValue`).
```

yrodiere commented:
@gsmet should I do anything or is it ok to merge ?

yrodiere committed:
hsearch-es: Clarify the name of some unit tests

I had renamed the tests in 961d95b3b8e02c2c104f6ef2bf96d82f822580c8
because I misunderstood their purpose from the comment on IndexedEntity
and thought they were misnamed, but actually it's the comment that was
unclear and should have been fixed.

gsmet commented on extensions/hibernate-search-elasticsearch/runtime/src/main/java/io/quarkus/hibernate/search/elasticsearch/runtime/HibernateSearchElasticsearchRuntimeConfig.java:
```diff
@@ -128,7 +128,7 @@
          */
-        @ConfigItem
+        @ConfigItem(name = ConfigItem.PARENT)
         ElasticsearchIndexConfig indexDefaults;
```
Maybe we could rename that one to `defaultBackend`? That would be clearer IMHO.
